### DISCUSSION TITLE
Volume reconstruction from segmented US images

### DISCUSSION THREAD

**Discussion by AzilizGP** [PlusToolkit/PlusLib#796] [Q&A] [answered]
Hi everyone !
I have been using fCal successfully to calibrate and track my ultrasound probe and I am now working on the volume reconstruction using Plus 2.9.0.
I have recorded US images of a pelvis sawbone and have been able to reconstruct the corresponding 3D volume using the volume reconstruction application in command line. I segmented some pelvis landmarks on the reconstructed volume (using 3DSlicer and ITKsnap) but I am very not sure of my segmentation as the images of the reconstructed volume do not look at all like the US images I acquired.
So I tried to first segment the US images (using ITK snap and exporting the segmentation in .mha format) and then use the volume reconstruction application to reconstruct the 3D volume of the segmented landmarks using this command:
`> VolumeReconstructor.exe --config-file=path\to\L-AIIS-config.xml --source-seq-file=path\to\L-AIIS-segmentation.mha --output-volume-file=path\to\L-AIIS-seg-volume.mha --image-to-reference-transform=ImageToTracker
`
The problem is that I only obtain this warning message : 
`
|WARNING|000.019000| Kinds not found in file: path\to\L-AIIS-segmentation.mha. Treating the last dimension as time.| in E:\D\PTNPTeb\IGSIO\Source\SequenceIO\vtkIGSIOMetaImageSequenceIO.cxx(199)
`
 and no output volume is saved. Sadly I do not understand this warning and did not find any information about it.

Does anyone have an idea on how to fix this problem or on how one can reconstruct a 3D volume from segmented US images ?

Here is a link to the recorded US images, the segmented US images, the config and the log file : https://filesender.renater.fr/?s=download&token=0999dff6-91d2-4099-951c-70791c5d91a2 

Thanks !
Aziliz

**Comment by adamrankin** [comment]
Yes, ITK snap exported mha files do not indicate what the dimensions of the file indicate. When they are omitted, Plus treats the last dimension as time.

Edit your header so that this:
```
ObjectType = Image
NDims = 3
BinaryData = True
BinaryDataByteOrderMSB = False
CompressedData = False
TransformMatrix = 1 0 0 0 1 0 0 0 1
Offset = 0 0 0
CenterOfRotation = 0 0 0
AnatomicalOrientation = RAI
ElementSpacing = 1 1 1
DimSize = 512 512 90
ElementType = MET_USHORT
ElementDataFile = LOCAL
```

becomes this:
```
ObjectType = Image
NDims = 3
AnatomicalOrientation = RAI
BinaryData = True
BinaryDataByteOrderMSB = False
CenterOfRotation = 0 0 0
CompressedData = False
TransformMatrix = 1 0 0 0 1 0 0 0 1  
DimSize = 512 512 90       
Kinds = domain domain domain
ElementNumberOfChannels = 1
ElementSpacing = 1 1 1
ElementType = MET_USHORT
Offset = 0 0 0
TransformMatrix = 1 0 0 0 1 0 0 0 1
UltrasoundImageOrientation = MN
UltrasoundImageType = BRIGHTNESS
ElementDataFile = LOCAL
```

Let me know if that loads correctly

**Reply by AzilizGP** [reply]
Thanks for the help! I just tried it and now I have a new error on failing to get the ImageToTracker transform:
|ERROR|000.101000| Transform path not found from Image to Tracker coordinate system. Available transforms in the repository (including the inverse of these transforms): ImageToProbe (valid, persistent), PhantomToReference (valid, persistent), StylusTipToStylus (valid, persistent), TransducerOriginPixelToTransducerOrigin (valid, persistent)| in E:\D\PTNPTeb\IGSIO\Source\IGSIOCommon\vtkIGSIOTransformRepository.cxx(543)
|WARNING|000.116000| Failed to get transform 'ImageToTracker' from transform repository!| in E:\D\PTNPTeb\IGSIO\Source\VolumeReconstruction\vtkIGSIOVolumeReconstructor.cxx(406)
|ERROR|000.124000| Automatic volume extent computation failed, there were no valid ImageToTracker transform available in the whole sequence (probably wrong image or reference coordinate system was defined or all transforms were invalid)| in E:\D\PTNPTeb\IGSIO\Source\VolumeReconstruction\vtkIGSIOVolumeReconstructor.cxx(429)
|ERROR|000.134000| Failed to set output extent of volume!| in E:\D\PTNPTeb\PlusLib\src\PlusVolumeReconstruction\Tools\VolumeReconstructor.cxx(181)
I fear that some metadata got lost during the segmentation export with ITKsnap. Is it possible that I retrieve the transformation from the recorded US images and add them to the segmented images ?

**Reply by adamrankin** [reply]
Ah, yes, all of the metadata that Plus uses for time-based series is lost. I would recommend exploring Andras' suggestion by using 3D Slicer to segment.

**Comment by lassoan** [comment, marked as answer]
Volume reconstructor algorithm is now directly in 3D Slicer (Volume Reconstruction module, provided by SlicerIGSIO extension):

https://youtu.be/2vXeJxYFou4

This allows continuous volume reconstruction and visualization in 3D and slice views, which is not just reduces latency (compared reconstructing volume in PlusServer), but also makes testing and troubleshooting easier, because you'll immediately see how acquired frames contribute to the reconstructed volume.
